src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncTargetBlockSelector.scala

```diff
@@ -98,27 +107,41 @@ class FastSyncTargetBlockSelector(
     }
   }
 
-  def waitingForTargetBlock(peer: Peer, targetBlockNumber: BigInt, timeout: Cancellable): Receive =
+  def waitingForTargetBlock(
+                             peersToAsk: Set[PeerId],
+                             targetBlockNumber: BigInt,
+                             timeout: Cancellable,
+                             headers: Map[BlockHeader, Int]
+                           ): Receive =
     handleCommonMessages orElse {
-    case MessageFromPeer(blockHeaders: BlockHeaders, _) =>
-      timeout.cancel()
-      peerEventBus ! Unsubscribe(MessageClassifier(Set(BlockHeaders.code), PeerSelector.WithId(peer.id)))
+    case MessageFromPeer(blockHeaders: BlockHeaders, peerId) =>
+      peerEventBus ! Unsubscribe(MessageClassifier(Set(BlockHeaders.code), PeerSelector.WithId(peerId)))
+
+      val updatedPeersToAsk = peersToAsk - peerId
 
       val targetBlockHeaderOpt = blockHeaders.headers.find(header => header.number == targetBlockNumber)
       targetBlockHeaderOpt match {
+        case Some(targetBlockHeader) if updatedPeersToAsk.isEmpty =>
+          timeout.cancel()
+          val newValue = headers.find(_._1 == targetBlockHeader).map(_._2).getOrElse(1)
+          val mostPopularBlockHeader = headers.updated(targetBlockHeader, newValue).maxBy(_._2)._1
+          sendResponseAndCleanup(mostPopularBlockHeader)
         case Some(targetBlockHeader) =>
-          sendResponseAndCleanup(targetBlockHeader)
+          val newValue = headers.find(_._1 == targetBlockHeader).map(_._2).getOrElse(1)
+          val updatedHeaders = headers.updated(targetBlockHeader, newValue)
+          context become waitingForTargetBlock(updatedPeersToAsk, targetBlockNumber, timeout, updatedHeaders)
         case None =>
-          blacklist(peer.id, blacklistDuration, s"did not respond with target block header, blacklisting and scheduling retry in $startRetryInterval")
-          log.info("Target block header not received. Retrying in {}", startRetryInterval)
-          scheduleRetry(startRetryInterval)
-          context become idle
+          blacklist(peerId, blacklistDuration, "Did not respond with target block header, blacklisting")
+          context become waitingForTargetBlock(updatedPeersToAsk, targetBlockNumber, timeout, headers)
       }
 
     case TargetBlockTimeout =>
-      blacklist(peer.id, blacklistDuration, s"did not respond with target block header (timeout), blacklisting and scheduling retry in $startRetryInterval")
+      peersToAsk.foreach{
+        peerId =>
+          blacklist(peerId, blacklistDuration, "Did not respond with target block header (timeout), blacklisting")
```

Comment:
we should proably have a timeout per peer to blackilst only the one which did not provide us with block.

Reply:
It is done here because `peersToAsk` is updated after each response and contains only not responding peers